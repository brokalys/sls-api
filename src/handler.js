import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (process.env.STAGE !== 'prod' && process.env.NODE_ENV !== 'test') ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (process.env.NODE_ENV !== 'test') {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
