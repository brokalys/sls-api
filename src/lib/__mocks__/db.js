export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
