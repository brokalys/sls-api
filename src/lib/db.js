import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
