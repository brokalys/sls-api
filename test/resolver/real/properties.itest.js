import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import { authenticateAs } from 'helpers';
import db from 'db-config';

const { query } = createTestClient(server);

describe('Resolver: properties - real queries from other services', () => {
  beforeAll(async () => {
    await db.migrate.latest();
    await db.seed.run();
  });

  afterEach(jest.clearAllMocks);

  describe('crawler', () => {
    test('city24.lv', async () => {
      authenticateAs('slsCrawler', server);

      const response = await query({
        query: `
          query GetExistingClassifieds($foreignIds: [String]) {
            properties(
              filter: {
                source: { eq: "city24.lv" }
                foreign_id: { in: $foreignIds }
              }
              limit: null
            ) {
              results {
                foreign_id
              }
            }
          }
        `,
        variables: {
          foreignIds: ['id111'],
        },
      });

      expect(response).toMatchSnapshot();
    });

    test('inch.lv', async () => {
      authenticateAs('slsCrawler', server);

      const response = await query({
        query: `
          query GetClassifieds(
            $url_0: String!
            $foreign_id_0: String!
            $url_1: String!
            $foreign_id_1: String!
          ) {
            row_0: properties(
              filter: {
                source: { eq: "inch.lv" }
                url: { eq: $url_0 }
                foreign_id: { eq: $foreign_id_0 }
              }
            ) {
              summary {
                count
              }
            }
            row_1: properties(
              filter: {
                source: { eq: "inch.lv" }
                url: { eq: $url_1 }
                foreign_id: { eq: $foreign_id_1 }
              }
            ) {
              summary {
                count
              }
            }
          }
        `,
        variables: {
          url_0: 'https://inch.lv',
          foreign_id_0: '10001',
          url_1: 'https://inch.lv',
          foreign_id_1: '10002',
        },
      });

      expect(response).toMatchSnapshot();
    });

    test('mm.lv', async () => {
      authenticateAs('slsCrawler', server);

      const response = await query({
        query: `
          query GetExistingClassifieds($urls: [String]) {
            properties(
              filter: {
                source: { eq: "mm-lv" }
                url: { in: $urls }
              }
              limit: null
            ) {
              results {
                url
              }
            }
          }
        `,
        variables: {
          urls: ['https://mm.lv'],
        },
      });

      expect(response).toMatchSnapshot();
    });

    test('ss.lv', async () => {
      authenticateAs('slsCrawler', server);

      const response = await query({
        query: `
          query GetExistingClassifieds(
            $source: String!
            $urls: [String!]!
            $created_at: String!
          ) {
            properties(
              filter: {
                source: { eq: $source }
                url: { in: $urls }
                created_at: { gte: $created_at }
              }
              limit: null
            ) {
              results {
                url
              }
            }
          }
        `,
        variables: {
          source: 'ss.lv',
          urls: [
            'https://www.ss.com/msg/lv/real-estate/flats/aizkraukle-and-reg/aiviekstes-pag/aghfx.html',
            'https://www.ss.lv/msg/lv/real-estate/flats/riga/bolderaya/cceeb.html',
          ],
          created_at: '2020-01-01T00:00:00.000Z',
        },
      });

      expect(response).toMatchSnapshot();
    });
  });

  describe('data extraction', () => {
    test('base-data', async () => {
      authenticateAs('slsDataExtraction', server);

      const response = await query({
        query: `
          query Extract(
            $created_at_start: String!
            $created_at_end: String!
            $type: String!
          ) {
            properties(
              filter: {
                created_at: { gte: $created_at_start, lte: $created_at_end }
                type: { eq: $type }
                price: { gte: 1 }
              },
              limit: null
            ) {
              summary {
                count
                price {
                  min
                  max
                  mean
                  median
                  mode
                  standardDev
                }
              }
            }
          }
        `,
        variables: {
          created_at_start: '2010-01-01T00:00:00.000Z',
          created_at_end: '2021-01-01T00:00:00.000Z',
          type: 'rent',
        },
      });

      expect(response).toMatchSnapshot();
    });

    test('regional-data', async () => {
      authenticateAs('slsDataExtraction', server);

      const response = await query({
        query: `
          query Extract(
            $created_at_start: String!
            $created_at_end: String!
            $category: String!
            $type: String!
          ) {
            properties(
              filter: {
                created_at: { gte: $created_at_start, lte: $created_at_end }
                category: { eq: $category }
                type: { eq: $type }
                rent_type: { in: ["monthly", "unknown"] }
                price: { gte: 1 }
              },
              limit: null
            ) {
              results {
                price
                lat
                lng
              }
            }
          }
        `,
        variables: {
          created_at_start: '2010-01-01T00:00:00.000Z',
          created_at_end: '2022-01-01T00:00:00.000Z',
          category: 'apartment',
          type: 'rent',
        },
      });

      expect(response).toMatchSnapshot();
    });

    test('regional-price-per-sqm', async () => {
      authenticateAs('slsDataExtraction', server);

      const response = await query({
        query: `
          query Extract(
            $created_at_start: String!
            $created_at_end: String!
            $category: String!
            $type: String!
          ) {
            properties(
              filter: {
                created_at: { gte: $created_at_start, lte: $created_at_end }
                category: { eq: $category }
                type: { eq: $type }
                rent_type: { in: ["monthly", "unknown"] }
                calc_price_per_sqm: { gt: 0 }
              },
              limit: null
            ) {
              results {
                calc_price_per_sqm
                lat
                lng
              }
            }
          }
        `,
        variables: {
          created_at_start: '2010-01-01T00:00:00.000Z',
          created_at_end: '2022-01-01T00:00:00.000Z',
          category: 'apartment',
          type: 'rent',
        },
      });

      expect(response).toMatchSnapshot();
    });
  });

  describe('static api', () => {
    test('regional-price-per-sqm', async () => {
      authenticateAs('slsStaticApi', server);

      const response = await query({
        query: `
          query Retrieve(
            $filter: PropertyFilter
          ) {
            properties(
              filter: $filter,
              limit: null
            ) {
              results {
                price
                price_per_sqm
              }
            }
          }
        `,
        variables: {
          filter: {
            published_at: {
              gte: '2010-01-01T00:00:00.000Z',
              lte: '2022-01-01T00:00:00.000Z',
            },
            category: {
              eq: 'apartment',
            },
            type: {
              eq: 'rent',
            },
            rent_type: {
              in: ['monthly', 'unknown'],
            },
            location_classificator: {
              eq: 'latvia-riga-agenskalns',
            },
          },
        },
      });

      expect(response).toMatchSnapshot();
    });
  });
});
