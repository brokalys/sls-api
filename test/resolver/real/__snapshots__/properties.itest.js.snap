// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolver: properties - real queries from other services crawler city24.lv 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "foreign_id": "id111",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: properties - real queries from other services crawler inch.lv 1`] = `
Object {
  "data": Object {
    "row_0": Object {
      "summary": Object {
        "count": 1,
      },
    },
    "row_1": Object {
      "summary": Object {
        "count": 0,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: properties - real queries from other services crawler mm.lv 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "url": "https://mm.lv",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: properties - real queries from other services crawler ss.lv 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "url": "https://www.ss.com/msg/lv/real-estate/flats/aizkraukle-and-reg/aiviekstes-pag/aghfx.html",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: properties - real queries from other services data extraction base-data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "summary": Object {
        "count": 2,
        "price": Object {
          "max": 300,
          "mean": 250,
          "median": 250,
          "min": 200,
          "mode": 200,
          "standardDev": 50,
        },
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: properties - real queries from other services data extraction regional-data 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "lat": 56.1,
          "lng": 24.9,
          "price": 100,
        },
        Object {
          "lat": 56.1,
          "lng": 24.9,
          "price": 200,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: properties - real queries from other services data extraction regional-price-per-sqm 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "area": 90,
          "lat": 56.1,
          "lng": 24.9,
          "price": 100,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Resolver: properties - real queries from other services static api regional-price-per-sqm 1`] = `
Object {
  "data": Object {
    "properties": Object {
      "results": Array [
        Object {
          "price": 100,
          "price_per_sqm": 1.11,
        },
        Object {
          "price": 200,
          "price_per_sqm": null,
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
